src/config-import.js

```diff
@@ -0,0 +1,231 @@
+import yauzl from 'yauzl-promise'
+import { validate, valueSchemas } from '@mapeo/schema'
+import { text, buffer } from 'node:stream/consumers'
+
+// Throw error if a zipfile contains more than 10,000 entries
+const MAX_ENTRIES = 10_000
+
+/**
+ * @typedef {Omit<yauzl.Entry, 'fileName'> & { filename: string }} Entry
+ */
+/**
+ * @typedef {{
+ *   presets: { [id: string]: unknown }
+ *   fields:  { [id: string]: unknown }
+ * }} PresetsFile
+ */
+/**
+ * @typedef {Parameters<import('./icon-api.js').IconApi['create']>[0]} IconData
+ */
+
+/**
+ * @param {string} configPath
+ */
+export async function readConfig(configPath) {
+  /** @type {Error[]} */
+  const warnings = []
+
+  const zip = await yauzl.open(configPath)
+  if (zip.entryCount > MAX_ENTRIES) {
+    throw new Error(`Zip file contains too many entries. Max is ${MAX_ENTRIES}`)
+  }
+  /** @type {Entry[]} */
+  const entries = /** @type {any} */ (await zip.readEntries(MAX_ENTRIES))
+
+  const presetsEntry = entries.find((e) => e.filename === 'presets.json')
+  if (!presetsEntry) {
+    throw new Error('Zip file does not contain presets.json')
+  }
+  const presetsFile = JSON.parse(
+    await text(await presetsEntry.openReadStream())
+  )
+  validatePresetsFile(presetsFile)
+
+  return {
+    get errors() {
+      if (warnings.length === 0) return null
+      return warnings
+    },
+
+    async close() {
+      zip.close()
+    },
+
+    /**
+     * @returns {AsyncIterable<IconData>}
+     */
+    async *icons() {
+      const iconEntries = entries
+        .filter((e) => e.filename.startsWith('/icons/'))
+        .sort()
+      /** @type {IconData | undefined} */
+      let icon
+      for (const entry of iconEntries) {
```

Comment:
If we want to avoid reading all entries into memory:

```suggestion
      /** @type {IconData | undefined} */
      let icon
      for await (const entry of zip) {
        if (!entry.filename.startsWith('/icons/')) continue
```